### PR TITLE
Change keystore cipher algorithm

### LINKED ISSUE #24509: Vulnerability detection seems to be disabled or has a problem #24457
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.8.1  | Vulnerability Detection Feed | Manager | Installation Assistant | Ubuntu 22 |

## Description
After performing the Wazuh single-node installation using the installation assistant for the https://github.com/wazuh/wazuh/issues/24457 E2E UX test, I encountered the following error on the Wazuh dashboard when I tried viewing the vulnerability for the agents.

image

Troubleshooting further, I saw the following errors on the Wazuh manager

```console
root@ip-172-31-34-147:/home/ubuntu# cat /var/ossec/logs/ossec.log | grep -i -E "error|warn"
2024/07/08 18:30:24 indexer-connector: WARNING: No username and password found in the keystore, using default values.
2024/07/08 18:30:24 indexer-connector: WARNING: IndexerConnector initialization failed for index 'wazuh-states-vulnerabilities-ip-172-31-34-147', retrying until the connection is successful.
2024/07/08 18:31:15 indexer-connector: WARNING: No username and password found in the keystore, using default values.
2024/07/08 18:31:15 indexer-connector: WARNING: IndexerConnector initialization failed for index 'wazuh-states-vulnerabilities-ip-172-31-34-147', retrying until the connection is successful.
2024/07/08 18:35:03 wazuh-modulesd:content-updater: WARNING: The offsets download has been interrupted
2024/07/08 18:35:24 indexer-connector: WARNING: No username and password found in the keystore, using default values.
2024/07/08 18:35:24 indexer-connector: WARNING: IndexerConnector initialization failed for index 'wazuh-states-vulnerabilities-ip-172-31-34-147', retrying until the connection is successful.
2024/07/08 19:23:19 wazuh-remoted: WARNING: Agent key already in use: agent ID '001'
2024/07/08 19:38:30 wazuh-modulesd:vulnerability-scanner: ERROR: Error updating feed: [json.exception.parse_error.101] parse error at line 1, column 9517943: syntax error while parsing value - invalid string: missing closing quote; last read: '"/contain', trying to re-download the feed.
2024/07/08 19:38:30 indexer-connector: WARNING: Failed to sync agent '001' with the indexer.
2024/07/08 19:38:31 indexer-connector: WARNING: Failed to sync agent '002' with the indexer.
2024/07/08 19:38:32 indexer-connector: WARNING: Failed to sync agent '003' with the indexer.
2024/07/08 19:38:33 wazuh-modulesd:vulnerability-scanner: ERROR: Error updating feed: [json.exception.out_of_range.401] array index 19 is out of range, trying to re-download the feed.
2024/07/08 21:44:22 indexer-connector: WARNING: Failed to sync agent '004' with the indexer.
2024/07/08 21:44:22 indexer-connector: WARNING: Failed to sync agent '001' with the indexer.
2024/07/08 21:44:22 indexer-connector: WARNING: Failed to sync agent '002' with the indexer.
2024/07/08 21:44:22 indexer-connector: WARNING: Failed to sync agent '003' with the indexer.
2024/07/08 21:44:43 indexer-connector: WARNING: No username and password found in the keystore, using default values.
2024/07/08 21:44:43 indexer-connector: WARNING: IndexerConnector initialization failed for index 'wazuh-states-vulnerabilities-ip-172-31-34-147', retrying until the connection is successful.
```
Below is the Wazuh manager configuration

```console
root@ip-172-31-34-147:/home/ubuntu# cat  /var/ossec/etc/ossec.conf
<!--
  Wazuh - Manager - Default configuration for ubuntu 22.04
  More info at: https://documentation.wazuh.com
  Mailing list: https://groups.google.com/forum/#!forum/wazuh
-->

<ossec_config>
  <global>
    <jsonout_output>yes</jsonout_output>
    <alerts_log>yes</alerts_log>
    <logall>no</logall>
    <logall_json>no</logall_json>
    <email_notification>no</email_notification>
    <smtp_server>smtp.example.wazuh.com</smtp_server>
    <email_from>wazuh@example.wazuh.com</email_from>
    <email_to>recipient@example.wazuh.com</email_to>
    <email_maxperhour>12</email_maxperhour>
    <email_log_source>alerts.log</email_log_source>
    <agents_disconnection_time>10m</agents_disconnection_time>
    <agents_disconnection_alert_time>0</agents_disconnection_alert_time>
    <update_check>yes</update_check>
  </global>

  <alerts>
    <log_alert_level>3</log_alert_level>
    <email_alert_level>12</email_alert_level>
  </alerts>

  <!-- Choose between "plain", "json", or "plain,json" for the format of internal logs -->
  <logging>
    <log_format>plain</log_format>
  </logging>

  <remote>
    <connection>secure</connection>
    <port>1514</port>
    <protocol>tcp</protocol>
    <queue_size>131072</queue_size>
  </remote>

  <!-- Policy monitoring -->
  <rootcheck>
    <disabled>no</disabled>
    <check_files>yes</check_files>
    <check_trojans>yes</check_trojans>
    <check_dev>yes</check_dev>
    <check_sys>yes</check_sys>
    <check_pids>yes</check_pids>
    <check_ports>yes</check_ports>
    <check_if>yes</check_if>

    <!-- Frequency that rootcheck is executed - every 12 hours -->
    <frequency>43200</frequency>

    <rootkit_files>etc/rootcheck/rootkit_files.txt</rootkit_files>
    <rootkit_trojans>etc/rootcheck/rootkit_trojans.txt</rootkit_trojans>

    <skip_nfs>yes</skip_nfs>
  </rootcheck>

  <wodle name="cis-cat">
    <disabled>yes</disabled>
    <timeout>1800</timeout>
    <interval>1d</interval>
    <scan-on-start>yes</scan-on-start>

    <java_path>wodles/java</java_path>
    <ciscat_path>wodles/ciscat</ciscat_path>
  </wodle>

  <!-- Osquery integration -->
  <wodle name="osquery">
    <disabled>yes</disabled>
    <run_daemon>yes</run_daemon>
    <log_path>/var/log/osquery/osqueryd.results.log</log_path>
    <config_path>/etc/osquery/osquery.conf</config_path>
    <add_labels>yes</add_labels>
  </wodle>

  <!-- System inventory -->
  <wodle name="syscollector">
    <disabled>no</disabled>
    <interval>1h</interval>
    <scan_on_start>yes</scan_on_start>
    <hardware>yes</hardware>
    <os>yes</os>
    <network>yes</network>
    <packages>yes</packages>
    <ports all="no">yes</ports>
    <processes>yes</processes>

    <!-- Database synchronization settings -->
    <synchronization>
      <max_eps>10</max_eps>
    </synchronization>
  </wodle>

  <sca>
    <enabled>yes</enabled>
    <scan_on_start>yes</scan_on_start>
    <interval>12h</interval>
    <skip_nfs>yes</skip_nfs>
  </sca>

  <vulnerability-detection>
    <enabled>yes</enabled>
    <index-status>yes</index-status>
    <feed-update-interval>60m</feed-update-interval>
  </vulnerability-detection>

  <indexer>
    <enabled>yes</enabled>
    <hosts>
      <host>https://127.0.0.1:9200</host>
    </hosts>
    <ssl>
      <certificate_authorities>
        <ca>/etc/filebeat/certs/root-ca.pem</ca>
      </certificate_authorities>
      <certificate>/etc/filebeat/certs/wazuh-1.pem</certificate>
      <key>/etc/filebeat/certs/wazuh-1-key.pem</key>
    </ssl>
  </indexer>

  <!-- File integrity monitoring -->
  <syscheck>
    <disabled>no</disabled>

    <!-- Frequency that syscheck is executed default every 12 hours -->
    <frequency>43200</frequency>

    <scan_on_start>yes</scan_on_start>

    <!-- Generate alert when new file detected -->
    <alert_new_files>yes</alert_new_files>

    <!-- Don't ignore files that change more than 'frequency' times -->
    <auto_ignore frequency="10" timeframe="3600">no</auto_ignore>

    <!-- Directories to check  (perform all possible verifications) -->
    <directories>/etc,/usr/bin,/usr/sbin</directories>
    <directories>/bin,/sbin,/boot</directories>

    <!-- Files/directories to ignore -->
    <ignore>/etc/mtab</ignore>
    <ignore>/etc/hosts.deny</ignore>
    <ignore>/etc/mail/statistics</ignore>
    <ignore>/etc/random-seed</ignore>
    <ignore>/etc/random.seed</ignore>
    <ignore>/etc/adjtime</ignore>
    <ignore>/etc/httpd/logs</ignore>
    <ignore>/etc/utmpx</ignore>
    <ignore>/etc/wtmpx</ignore>
    <ignore>/etc/cups/certs</ignore>
    <ignore>/etc/dumpdates</ignore>
    <ignore>/etc/svc/volatile</ignore>

    <!-- File types to ignore -->
    <ignore type="sregex">.log$|.swp$</ignore>

    <!-- Check the file, but never compute the diff -->
    <nodiff>/etc/ssl/private.key</nodiff>

    <skip_nfs>yes</skip_nfs>
    <skip_dev>yes</skip_dev>
    <skip_proc>yes</skip_proc>
    <skip_sys>yes</skip_sys>

    <!-- Nice value for Syscheck process -->
    <process_priority>10</process_priority>

    <!-- Maximum output throughput -->
    <max_eps>50</max_eps>

    <!-- Database synchronization settings -->
    <synchronization>
      <enabled>yes</enabled>
      <interval>5m</interval>
      <max_eps>10</max_eps>
    </synchronization>
  </syscheck>

  <!-- Active response -->
  <global>
    <white_list>127.0.0.1</white_list>
    <white_list>^localhost.localdomain$</white_list>
    <white_list>127.0.0.53</white_list>
  </global>

  <command>
    <name>disable-account</name>
    <executable>disable-account</executable>
    <timeout_allowed>yes</timeout_allowed>
  </command>

  <command>
    <name>restart-wazuh</name>
    <executable>restart-wazuh</executable>
  </command>

  <command>
    <name>firewall-drop</name>
    <executable>firewall-drop</executable>
    <timeout_allowed>yes</timeout_allowed>
  </command>

  <command>
    <name>host-deny</name>
    <executable>host-deny</executable>
    <timeout_allowed>yes</timeout_allowed>
  </command>

  <command>
    <name>route-null</name>
    <executable>route-null</executable>
    <timeout_allowed>yes</timeout_allowed>
  </command>

  <command>
    <name>win_route-null</name>
    <executable>route-null.exe</executable>
    <timeout_allowed>yes</timeout_allowed>
  </command>

  <command>
    <name>netsh</name>
    <executable>netsh.exe</executable>
    <timeout_allowed>yes</timeout_allowed>
  </command>

  <!--
  <active-response>
    active-response options here
  </active-response>
  -->

  <!-- Log analysis -->
  <localfile>
    <log_format>command</log_format>
    <command>df -P</command>
    <frequency>360</frequency>
  </localfile>

  <localfile>
    <log_format>full_command</log_format>
    <command>netstat -tulpn | sed 's/\([[:alnum:]]\+\)\ \+[[:digit:]]\+\ \+[[:digit:]]\+\ \+\(.*\):\([[:digit:]]*\)\ \+\([0-9\.\:\*]\+\).\+\ \([[:digit:]]*\/[[:alnum:]\-]*\).*/\1 \2 == \3 == \4 \5/' | sort -k 4 -g | sed 's/ == \(.*\) ==/:\1/' | sed 1,2d</command>
    <alias>netstat listening ports</alias>
    <frequency>360</frequency>
  </localfile>

  <localfile>
    <log_format>full_command</log_format>
    <command>last -n 20</command>
    <frequency>360</frequency>
  </localfile>

  <ruleset>
    <!-- Default ruleset -->
    <decoder_dir>ruleset/decoders</decoder_dir>
    <rule_dir>ruleset/rules</rule_dir>
    <rule_exclude>0215-policy_rules.xml</rule_exclude>
    <list>etc/lists/audit-keys</list>
    <list>etc/lists/amazon/aws-eventnames</list>
    <list>etc/lists/security-eventchannel</list>

    <!-- User-defined ruleset -->
    <decoder_dir>etc/decoders</decoder_dir>
    <rule_dir>etc/rules</rule_dir>
  </ruleset>

  <rule_test>
    <enabled>yes</enabled>
    <threads>1</threads>
    <max_sessions>64</max_sessions>
    <session_timeout>15m</session_timeout>
  </rule_test>

  <!-- Configuration for wazuh-authd -->
  <auth>
    <disabled>no</disabled>
    <port>1515</port>
    <use_source_ip>no</use_source_ip>
    <purge>yes</purge>
    <use_password>no</use_password>
    <ciphers>HIGH:!ADH:!EXP:!MD5:!RC4:!3DES:!CAMELLIA:@STRENGTH</ciphers>
    <!-- <ssl_agent_ca></ssl_agent_ca> -->
    <ssl_verify_host>no</ssl_verify_host>
    <ssl_manager_cert>etc/sslmanager.cert</ssl_manager_cert>
    <ssl_manager_key>etc/sslmanager.key</ssl_manager_key>
    <ssl_auto_negotiate>no</ssl_auto_negotiate>
  </auth>

  <cluster>
    <name>wazuh</name>
    <node_name>node01</node_name>
    <node_type>master</node_type>
    <key></key>
    <port>1516</port>
    <bind_addr>0.0.0.0</bind_addr>
    <nodes>
        <node>NODE_IP</node>
    </nodes>
    <hidden>no</hidden>
    <disabled>yes</disabled>
  </cluster>

</ossec_config>

<ossec_config>
  <localfile>
    <log_format>syslog</log_format>
    <location>/var/ossec/logs/active-responses.log</location>
  </localfile>

  <localfile>
    <log_format>syslog</log_format>
    <location>/var/log/auth.log</location>
  </localfile>

  <localfile>
    <log_format>syslog</log_format>
    <location>/var/log/syslog</location>
  </localfile>

  <localfile>
    <log_format>syslog</log_format>
    <location>/var/log/dpkg.log</location>
  </localfile>

  <localfile>
    <log_format>syslog</log_format>
    <location>/var/log/kern.log</location>
  </localfile>

</ossec_config>

```
I did further troubleshooting by checking the Filebeat configuration file and adding the credentials of the Wazuh indexer to the Wazuh manager keystore. The error persisted after performing all these steps.

```console
root@ip-172-31-34-147:/home/ubuntu# cat /etc/filebeat/filebeat.yml
# Wazuh - Filebeat configuration file
output.elasticsearch:
  protocol: https
  username: ${username}
  password: ${password}
  ssl.certificate_authorities:
    - /etc/filebeat/certs/root-ca.pem
  ssl.certificate: "/etc/filebeat/certs/wazuh-1.pem"
  ssl.key: "/etc/filebeat/certs/wazuh-1-key.pem"
setup.template.json.enabled: true
setup.template.json.path: '/etc/filebeat/wazuh-template.json'
setup.template.json.name: 'wazuh'
setup.ilm.overwrite: true
setup.ilm.enabled: false

filebeat.modules:
  - module: wazuh
    alerts:
      enabled: true
    archives:
      enabled: false

logging.level: info
logging.to_files: true
logging.files:
  path: /var/log/filebeat
  name: filebeat
  keepfiles: 7
  permissions: 0644

logging.metrics.enabled: false

seccomp:
  default_action: allow
  syscalls:
  - action: allow
    names:
    - rseq

output.elasticsearch.hosts:
  - 127.0.0.1:9200
root@ip-172-31-34-147:/home/ubuntu# openssl x509 -in /etc/filebeat/certs/wazuh-1.pem -text -noout | grep IP
                IP Address:127.0.0.1
root@ip-172-31-34-147:/home/ubuntu# ls /etc/filebeat/certs
root-ca.pem  wazuh-1-key.pem  wazuh-1.pem
root@ip-172-31-34-147:/home/ubuntu# /var/ossec/bin/wazuh-keystore -f indexer -k username -v admin
root@ip-172-31-34-147:/home/ubuntu# /var/ossec/bin/wazuh-keystore -f indexer -k password -v P.ie2eZI5S.9FLPLSRB9j7aBINgBabl.
```console

root@ip-172-31-34-147:/home/ubuntu# /var/ossec/bin/wazuh-keystore -f indexer -k username -v admin
root@ip-172-31-34-147:/home/ubuntu# /var/ossec/bin/wazuh-keystore -f indexer -k password -v P.ie2eZI5S.9FLPLSRB9j7aBINgBabl.
root@ip-172-31-34-147:/home/ubuntu# systemctl restart wazuh-indexer
root@ip-172-31-34-147:/home/ubuntu# systemctl restart wazuh-manager
root@ip-172-31-34-147:/home/ubuntu# curl -u admin:P.ie2eZI5S.9FLPLSRB9j7aBINgBabl. --cacert /etc/filebeat/certs/root-ca.pem --cert /etc/filebeat/certs/wazuh-1.pem --key /etc/filebeat/certs/wazuh-1-key.pem -X GET "https://127.0.0.1:9200/_cluster/health"
{"cluster_name":"wazuh-indexer-cluster","status":"green","timed_out":false,"number_of_nodes":1,"number_of_data_nodes":1,"discovered_master":true,"discovered_cluster_manager":true,"active_primary_shards":13,"active_shards":13,"relocating_shards":0,"initializing_shards":0,"unassigned_shards":0,"delayed_unassigned_shards":0,"number_of_pending_tasks":0,"number_of_in_flight_fetch":0,"task_max_waiting_in_queue_millis":0,"active_shards_percent_as_number":100.0}

root@ip-172-31-34-147:/home/ubuntu# cat /var/ossec/logs/ossec.log | grep -i -E "error|warn"
2024/07/09 01:04:34 indexer-connector: WARNING: Failed to sync agent '002' with the indexer.
2024/07/09 01:04:34 indexer-connector: WARNING: Failed to sync agent '003' with the indexer.
2024/07/09 01:04:34 indexer-connector: WARNING: Failed to sync agent '004' with the indexer.
2024/07/09 01:04:34 indexer-connector: WARNING: Failed to sync agent '001' with the indexer.
2024/07/09 01:08:22 indexer-connector: WARNING: No username and password found in the keystore, using default values.
2024/07/09 01:08:23 indexer-connector: WARNING: IndexerConnector initialization failed for index 'wazuh-states-vulnerabilities-ip-172-31-34-147', retrying until the connection is successful.
root@ip-172-31-34-147:/home/ubuntu#
```

### PR DESCRIPTION
**Pull request by Dwordcito**
|Related issue|
|---|
|Close #24509|

This PR aims to remove the reuse of the sslmanager.cert and sslmanager.key.

We cipher the values in the keystore based on random numbers and using aes256 cbc.

An upgrade mechanism is introduced to avoid steps during the movement from the old algorithm to this new algorithm.

### CODE REVIEW

File: src/shared_modules/keystore/src/keyStore.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    {
+        try
+        {
+            // Check if the private key file exists
+            if (!std::filesystem::exists(PRIVATE_KEY_FILE))
+            {
+                throw std::runtime_error("Private key file not found.");
+            }
+
+            // Upgrade all keys
```
Comment (sebasfalcone) [inline src/shared_modules/keystore/src/keyStore.cpp:37]: Nitpick: rawValue can have a smaller scope
```suggestion

    // If the version field does not exist, it means that the keystore has not been upgraded yet.
    if (!keystoreDB.get(KS_VERSION_FIELD, versionValue, columnFamily))
    {
        try
        {
            // Check if the private key file exists
            if (!std::filesystem::exists(PRIVATE_KEY_FILE))
            {
                throw std::runtime_error("Private key file not found.");
            }

            std::string rawValue;
            // Upgrade all keys
```

File: src/shared_modules/keystore/src/keyStore.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+                // Insert the key-value pair using AES encryption
+                keystoreDB.put(key, rocksdb::Slice(encryptedValue.data(), encryptedValue.size()), columnFamily);
+            }
+        }
+        catch (const std::exception& exception)
+        {
+            // If the upgrade fails, delete all keys and log the error.
+            keystoreDB.deleteAll(columnFamily);
+            logWarn(KS_NAME, "Keystore upgrade failed, re-run this tool with all keys. Error: %s", exception.what());
```
Comment (sebasfalcone) [inline src/shared_modules/keystore/src/keyStore.cpp:63]: This warning doesn't appear if we run the keystore binary. We may want to print this to the cerr

File: src/shared_modules/utils/evpHelper.hpp [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    {
+        EVP_CIPHER_CTX* ctx;
+        int len;
+        int ciphertextLen;
+        std::vector<unsigned char> ciphertext(input.length() + T::AES_BLOCK_LENGTH, 0);
+        std::array<unsigned char, CIPHER_KEY_SIZE> key {};
+        std::array<unsigned char, CIPHER_IV_SIZE> iv {};
+
+        T::RAND_bytes(key.data(), key.size());
+        T::RAND_bytes(iv.data(), iv.size());
```
Comment (sebasfalcone) [inline src/shared_modules/utils/evpHelper.hpp:50]: We are not validating the return value of this method
```C++
int RAND_bytes_ex(OSSL_LIB_CTX *ctx, unsigned char *buf, size_t num,
                  unsigned int strength)
{
    EVP_RAND_CTX *rand;
#if !defined(OPENSSL_NO_DEPRECATED_3_0) && !defined(FIPS_MODULE)
    const RAND_METHOD *meth = RAND_get_rand_method();

    if (meth != NULL && meth != RAND_OpenSSL()) {
        if (meth->bytes != NULL)
            return meth->bytes(buf, num);
        ERR_raise(ERR_LIB_RAND, RAND_R_FUNC_NOT_IMPLEMENTED);
        return -1;
    }
#endif

    rand = RAND_get0_public(ctx);
    if (rand != NULL)
        return EVP_RAND_generate(rand, buf, num, strength, 0, NULL, 0);

    return 0;
}
```

**Review by sebasfalcone — COMMENTED** [review]
## Code 🟡 
- Some minor nitpicks that don't affect the code behavior
- Overall clean and understandable code

## Testing 🟢 

### Upgrade test - Normal functionality

Save value with keystore V1
```bash
root@Thinkpad-Sebas:/var/ossec/bin# ./wazuh-keystore-old -k key -v value -f column
Previous value: 
```

Update value with keystore V1
```bash
root@Thinkpad-Sebas:/var/ossec/bin# ./wazuh-keystore-old -k key -v value1 -f column
Previous value: value
```

Update value with keystore V2. We can decrypt previous values
```bash
root@Thinkpad-Sebas:/var/ossec/bin# ./wazuh-keystore -k key -v value2 -f column
Previous value: value1
```

Version key is added
```bash
./rocks_db_query_testtool -d /var/ossec/queue/keystore -k version -c column
version ==> 2
```

Save value with keystore V1. We are unable to decrypt the previous value
```bash
root@Thinkpad-Sebas:/var/ossec/bin# ./wazuh-keystore-old -k key -v value3 -f column
RSA decryption failed
```

### Upgrade test - No certificates

Save three values using keystore V1
```bash
./rocks_db_query_testtool -d /var/ossec/queue/keystore -k key -c column
key ==> ENCRYPTED DATA

./rocks_db_query_testtool -d /var/ossec/queue/keystore -k key2 -c column
key2 ==> ENCRYPTED DATA

./rocks_db_query_testtool -d /var/ossec/queue/keystore -k key3 -c column
key3 ==> ENCRYPTED DATA                                                                                                                                                        ```
```

Remove the certificates from the etc directory
```bash
mv ../etc/sslmanager.cert ../etc/sslmanager.cert.old && mv ../etc/sslmanager.key ../etc/sslmanager.key.old
```

Update value using the keystore V2 (No previous value was read because the column was deleted)
```bash
./wazuh-keystore -k key -v value -f column
Previous value: 
```

Old data was deleted and new data was saved
```bash
./rocks_db_query_testtool -d /var/ossec/queue/keystore -k key -c column
key ==> ENCRYPTED DATA 

./rocks_db_query_testtool -d /var/ossec/queue/keystore -k key3 -c column
Unable to find resource.

./rocks_db_query_testtool -d /var/ossec/queue/keystore -k key3 -c column
Unable to find resource.
```

The DB is operational despite the error, new values were saved

File: src/shared_modules/keystore/src/keyStore.cpp (hunk 2 above)
Comment (Dwordcito) [reply src/shared_modules/keystore/src/keyStore.cpp:63]: This warning appears when the method "Get" from the keystone class is called, and the indexer connector use this get method.

Comment (sebasfalcone) [reply src/shared_modules/keystore/src/keyStore.cpp:64]: It is part of the upgrade process, I am concerned about the behavior from the user's perspective

1. Have data stored using the old keystore in the column "a"
2. Certs are wrong for some reason
2. Use the new keystore in the column "a"
3. All the other values will be removed and the user was not warned

image

> [!NOTE]
> I know that once the certs are lost, there is no way to recover them but te user can't tell that this happen

File: src/shared_modules/utils/evpHelper.hpp [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+template<typename T = OpenSSLPrimitives>
+class EVPHelper final : public T
+{
+public:
+    explicit EVPHelper() = default;
+    ~EVPHelper() override = default;
+
+    /**
+     * Encrypts the input vector with the provided key
```
Comment (sebasfalcone) [inline src/shared_modules/utils/evpHelper.hpp:34]: We are not providing any key
```suggestion
     * Encrypts the input vector
```

File: src/shared_modules/utils/evpHelper.hpp [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+        // Save the key + iv + ciphertext in the output
+        output.resize(key.size() + iv.size() + ciphertextLen);
+        std::copy(key.begin(), key.end(), output.begin());
+        std::copy(iv.begin(), iv.end(), output.begin() + key.size());
+        std::copy(ciphertext.begin(), ciphertext.begin() + ciphertextLen, output.begin() + key.size() + iv.size());
+    }
+
+    /**
+     * Decrypts the input vector with the provided key
```
Comment (sebasfalcone) [inline src/shared_modules/utils/evpHelper.hpp:92]: We are not providing any key
```suggestion
     * Decrypts the input vector
```

Maybe we can clarify that the key is retrieved from the input vector

File: src/shared_modules/keystore/src/keyStore.cpp (hunk 2 above)
Comment (Dwordcito) [reply src/shared_modules/keystore/src/keyStore.cpp:64]: Your point makes sense, I'll add it.

File: src/shared_modules/keystore/src/main.cpp [hunk 6]
```diff
@@ -27,6 +29,28 @@ int main(int argc, char* argv[])
...
+           const std::string&,
+           const int,
+           const std::string&,
+           const std::string& str,
+           va_list args)
+        {
+            char formattedStr[MAXLEN] = {0};
+            vsnprintf(formattedStr, MAXLEN, str.c_str(), args);
+
+            if (logLevel == Log::LOGLEVEL_ERROR)
```
Comment (sebasfalcone) [inline src/shared_modules/keystore/src/main.cpp:44]:
```suggestion
            if (logLevel == Log::LOGLEVEL_ERROR || logLevel == Log::LOGLEVEL_WARNING ||
                logLevel == Log::LOGLEVEL_CRITICAL)
```

File: src/shared_modules/keystore/tests/component/main.cpp [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+           const std::string&,
+           const int,
+           const std::string&,
+           const std::string& str,
+           va_list args)
+        {
+            char formattedStr[MAXLEN] = {0};
+            vsnprintf(formattedStr, MAXLEN, str.c_str(), args);
+
+            if (logLevel == Log::LOGLEVEL_ERROR)
```
Comment (sebasfalcone) [inline src/shared_modules/keystore/tests/component/main.cpp:36]:
```suggestion
            if (logLevel == Log::LOGLEVEL_ERROR || logLevel == Log::LOGLEVEL_WARNING ||
                logLevel == Log::LOGLEVEL_CRITICAL)```

File: src/shared_modules/keystore/src/keyStore.cpp [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+                logInfo(KS_NAME, "Key pair '%s' upgraded.", key.c_str());
+            }
+        }
+        catch (const std::exception& exception)
+        {
+            // If the upgrade fails, delete all keys and log the error.
+            keystoreDB.deleteAll(columnFamily);
+            logWarn(KS_NAME,
+                    "Keystore upgrade failed, re-run the tool again for all keys to save. Error: %s",
+                    exception.what());
```
Comment (sebasfalcone) [inline src/shared_modules/keystore/src/keyStore.cpp:68]: The message is not clear enough

```suggestion
            logWarn(KS_NAME,
                    "Keystore upgrade failed. Re-run the tool to regenerate the deleted keys in column '%s'. Error: %s",
                    columnFamily,
                    exception.what());
```

OR: 

```suggestion
            logWarn(KS_NAME,
                    "Keystore upgrade failed, re-run the tool again for all keys to save them. Error: %s",
                    columnFamily,
                    exception.what());```

**Review by sebasfalcone — COMMENTED** [review]
One last iteration

**Review by sebasfalcone — APPROVED** [review]
LGTM

**Pull request merged**